spec/controllers/session_controller_spec.rb

```diff
@@ -0,0 +1,38 @@
+require 'rails_helper'
+
+RSpec.describe Devise::SessionsController, type: :controller do
+  it 'Get to signin page' do
+    @request.env['devise.mapping'] = Devise.mappings[:user]
+    get :new
+    expect(response.status).to eq(200)
+  end
+
+  it 'Get 302 on User and Password request for TOTP' do
+    strategy = Devise::Strategies::RemoteAuthenticatable.new(nil)
+    allow(request).to receive(:headers).and_return(user: 'name')
+    SelfService.service(:cognito_client).stub_responses(:initiate_auth, challenge_name: 'SOFTWARE_TOKEN_MFA', session: SecureRandom.uuid, challenge_parameters: { 'FRIENDLY_DEVICE_NAME' => 'Authy' })
+    @request.env['devise.mapping'] = Devise.mappings[:user]
+    post :create, params: { user: { email: 'test@test.com', password: 'validpass' } }
+    expect(response.status).to eq(302)
+    expect(subject).to redirect_to(new_user_session_path)
+  end
+
+  it 'Get 302 to root path on TOTP request' do
+    strategy = Devise::Strategies::RemoteAuthenticatable.new(nil)
+    SelfService.service(:cognito_client).stub_responses(:respond_to_auth_challenge, challenge_name: nil, authentication_result: { access_token: 'valid-token' })
+    allow(request).to receive(:headers).and_return(user: 'name')
+    allow(strategy).to receive(:params).at_least(:once).and_return(user: 'name')
+    session[:challenge_name] = 'MFA'
+    session[:cognito_session_id] = SecureRandom.uuid
+    session[:challenge_parameters] = { 'FRIENDLY_DEVICE_NAME' => 'Authy' }
+    session[:username] = 'test@test.com'
+    @request.env['devise.mapping'] = Devise.mappings[:user]
+    post :create, params: { user: { email: 'test@test.com', totp_code: '999999' } }
+    expect(response.status).to eq(302)
+    expect(subject).to redirect_to(root_path)
+    expect(session.key?('challenge_name')).to eq(false)
```

Comment:
Because our logic looks at if we actually have the key... I think its better to check we have actually removed the key rather than just checking if the response is nil.  So I'd like not to change this if thats OK?  Otherwise we should also change our logic as well as the test.

Reply:
it's not checking any responses, but accessing the session directly. It's doing the same thing like yours, except your one is calling a method on the session and then doing comparison.